From 4230e63e97547526bccfb74ea7a91d4136f9476c Mon Sep 17 00:00:00 2001
From: Yuta Saito <kateinoigakukun@gmail.com>
Date: Sat, 4 May 2024 13:09:57 +0000
Subject: [PATCH 1/2] getdtablesize: support systems without getrlimit like
 WASI

* lib/getdtablesize.c (getdtablesize): Return INT_MAX if getrlimit
  is not available. This is the case for WASI, WebAssembly System
  Interface.
* m4/getdtablesize.m4 (gl_PREREQ_GETDTABLESIZE): Check for getrlimit.
---
 lib/getdtablesize.c | 2 ++
 m4/getdtablesize.m4 | 7 +++++--
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/lib/getdtablesize.c b/lib/getdtablesize.c
index 762c100b38..fca937a39a 100644
--- a/lib/getdtablesize.c
+++ b/lib/getdtablesize.c
@@ -109,6 +109,7 @@ getdtablesize (void)
 int
 getdtablesize (void)
 {
+# if HAVE_GETRLIMIT
   struct rlimit lim;
 
   if (getrlimit (RLIMIT_NOFILE, &lim) == 0
@@ -117,6 +118,7 @@ getdtablesize (void)
       && lim.rlim_cur != RLIM_SAVED_CUR
       && lim.rlim_cur != RLIM_SAVED_MAX)
     return lim.rlim_cur;
+# endif
 
   return INT_MAX;
 }
diff --git a/m4/getdtablesize.m4 b/m4/getdtablesize.m4
index 3b89456baf..7059eb933e 100644
--- a/m4/getdtablesize.m4
+++ b/m4/getdtablesize.m4
@@ -1,4 +1,5 @@
-# getdtablesize.m4 serial 8
+# getdtablesize.m4
+# serial 9
 dnl Copyright (C) 2008-2024 Free Software Foundation, Inc.
 dnl This file is free software; the Free Software Foundation
 dnl gives unlimited permission to copy and/or distribute it,
@@ -60,4 +61,6 @@ AC_DEFUN([gl_FUNC_GETDTABLESIZE],
 ])
 
 # Prerequisites of lib/getdtablesize.c.
-AC_DEFUN([gl_PREREQ_GETDTABLESIZE], [:])
+AC_DEFUN([gl_PREREQ_GETDTABLESIZE], [
+  AC_CHECK_FUNCS_ONCE([getrlimit])
+])
-- 
2.43.2

